alacritty_config_derive/src/de_struct.rs

```diff
@@ -0,0 +1,218 @@
+            let parsed = match attr.parse_args::<Attr>() {
+                Ok(parsed) => parsed,
+                Err(_) => continue,
+            };
+
+            match parsed.ident.as_str() {
+                // Skip deserialization for `#[config(skip)]` fields.
+                "skip" => continue 'fields_loop,
+                "flatten" => {
+                    // NOTE: Currently only a single instance of flatten is supported per struct
+                    // for complexity reasons.
+                    if !field_streams.flatten.is_empty() {
+                        let error = Error::new(attr.span(), MULTIPLE_FLATTEN_ERROR);
+                        field_streams.flatten = error.to_compile_error();
+                        return field_streams;
+                    }
+
+                    // Create the tokens to deserialize the flattened struct from the unused fields.
+                    field_streams.flatten.extend(quote! {
+                        let unused = serde_yaml::Value::Mapping(unused);
+                        config.#ident = serde::Deserialize::deserialize(unused).unwrap_or_default();
+                    });
+                },
+                "deprecated" => {
+                    // Construct deprecation message and append optional attribute override.
+                    let mut message =
+                        format!("Config warning: {} is deprecated", ident.to_string());
+                    if let Some(warning) = parsed.param {
+                        message = format!("{}; {}", message, warning.value());
+                    }
+
+                    // Append stream to log deprecation warning.
+                    match_assignment_stream.extend(quote! {
+                        log::warn!(target: #LOG_TARGET, #message);
+                    });
+                },
+                // Add aliases to match pattern.
+                "alias" => {
+                    if let Some(alias) = parsed.param {
+                        literals.push(alias.value());
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        if let Type::Path(type_path) = &field.ty {
+            if crate::path_ends_with(&type_path.path, "Option") {
+                // Create token stream for deserializing "none" string into `Option<T>`.
+                match_assignment_stream = quote! {
+                    if value.as_str().map_or(false, |s| s.eq_ignore_ascii_case("none")) {
+                        config.#ident = None;
+                        continue;
+                    }
+                    #match_assignment_stream
+                };
+            }
+        }
+
+        // Create token stream for deserialization and error handling.
+        field_streams.match_assignments.extend(quote! {
+            #(#literals)|* => { #match_assignment_stream },
+        });
+    }
+
+    field_streams
+}
```

Comment:
This is an absolutely massive function and I've tried to refactor it a few times to see if there's maybe a more elegant way to do this, but I couldn't find anything that would actually make this easier to read.

Since there's a lot of references to shared data over the entire lifetime of that method, it makes it a bit difficult to move things into functions without creating unreasonable control flow or massive parameter lists.